### DISCUSSION TITLE
How to get the author of a PR?

### DISCUSSION THREAD

**Discussion by zachliu** [community/community#25502] [Actions]
<p>I have these in my yml file:</p>
<pre><code class="lang-auto">on:
  issue_comment:
    types: [created]
</code></pre>
<p>But since the one who issues the comment may not be the PR author, the <code>github.actor</code> is not the PR author’s username. Also, since there is no “pusher” here, <code>github.event.pusher.name</code> doesn’t exist. Is there a way to obtain the PR author in this case?</p>
<p>In other words, how do we get the PR author when a random person issues a comment?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<p>When a workflow run is triggered by the events related to pull request, from the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="noopener nofollow ugc"><strong>github context</strong></a> of this run you can see the “<strong>github.event.pull_request</strong>” object. In this object, you can use the property “<strong>github.event.pull_request.user.login</strong>” to get the author who opened this PR.<br>
For example:</p>
<pre><code class="lang-yaml">- name: View PR author
  run: echo "This PR is opened by ${{ github.event.pull_request.user.login }} ."
</code></pre>

**Reply by nitishgupta08** [reply]
This gives name of the owner of repository instead of the user who created/updated PR. Any idea how I can fix that

**Comment by zachliu** [comment]
<p>Thanks. But in my case, the workflow is triggered by issuing a comment. Will this still work?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<p>Sure. When you add a comments on a PR, the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#issue_comment" rel="noopener nofollow ugc"><strong>issue_comment</strong></a> event is triggered. At this time, the <strong>issue_comment</strong> event is related to the PR.</p>
<p><strong>Tip:</strong> On GitHub, pull request is considered as a type of issue, that’s way some event types of PR will be combined as the common issue events.</p>

**Comment by zachliu** [comment]
<p>Great! I’ll give it a go and report back :pray:</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,<br>
Okay. Please go ahead, and any update, feel free to tell me.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<p><strong>EDIT:</strong></p>
<p>When the workflow is triggered by  the <strong>issue_comment</strong> event, the object will always be “<strong>github.event.issue</strong>”, so the author should be “<strong>github.event.issue.user.login</strong>”.</p>
<p>To determine if it is a PR, you can see if the “<strong>github.event.issue.pull_request</strong>” object exists.<br>
for example:</p>
<pre><code class="lang-yaml">- name: View PR author
  if: ${{ github.event.issue.pull_request }}
  run: echo "This PR is opened by ${{ github.event.issue.user.login }} ."
</code></pre>

**Comment by zachliu** [comment]
<p>It’s a bit confusing now. Let me tell the story again:<br>
If A opens a PR and triggers a workflow, it’s easy to get A’s username in the workflow<br>
If B comments on A’s PR and triggers another workflow, how do I still get A’s username in this workflow?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<blockquote>
<p>If A opens a PR and triggers a workflow,</p>
</blockquote>
<p>At this time, the workflow is triggered by <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request" rel="noopener nofollow ugc"><strong>pull_request</strong></a> event, to get the author who opened the PR, using the expression “<strong>github.event.pull_request.user.login</strong>”.<br>
The events “<a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_review" rel="noopener nofollow ugc"><strong>pull_request_review</strong></a>”, “<a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_review_comment" rel="noopener nofollow ugc"><strong>pull_request_review_comment</strong></a>” and “<a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#pull_request_target" rel="noopener nofollow ugc"><strong>pull_request_target</strong></a>” use the same expression.</p>
<br>
<blockquote>
<p>If B comments on A’s PR and triggers another workflow</p>
</blockquote>
<p>At this time, the workflow is triggered by <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#issue_comment" rel="noopener nofollow ugc"><strong>issue_comment</strong></a> event, to get the author who opened the PR, using the expression “<strong>github.event.issue.user.login</strong>”.</p>

**Comment by zachliu** [comment]
<p>Is it <code>issue</code> or <code>issues</code> in <code>{{ github.event.issue.user.login }}</code>? You meant <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#issues" rel="noopener nofollow ugc">issues</a> right?<br>
I can’t find <code>issue</code> under <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#webhook-events" rel="noopener nofollow ugc">webhook events</a>.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<p>No, I’m just saying the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#issue_comment" rel="noopener nofollow ugc"><strong>issue_comment</strong></a> event, not <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#issues" rel="noopener nofollow ugc"><strong>issues</strong></a> event.</p>
<p>As I mentioned above, pull request is considered as a type of issue, but a PR is not exactly same with an issue.</p>
<ul>
<li>
<p>There are some event types are the communal that can be used for both <strong>Pull requests</strong> and <strong>Issues</strong>, such as the <strong>issue_comment</strong> event.</p>
</li>
<li>
<p>However, some event types only can be used for <strong>Pull requests</strong>, such as <strong>pull_request</strong>, <strong>pull_request_review</strong>, <strong>pull_request_review_comment</strong> and <strong>pull_request_target</strong> events.</p>
</li>
<li>
<p>And some only for <strong>Issues</strong> (such as <strong>issues</strong> event).</p>
</li>
</ul>
<p>Every time a workflow run is triggered, GitHub will generate the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="noopener nofollow ugc"><strong>github context</strong></a> for this run. The <strong>github context</strong> contains many detailed information about the current workflow run, such as the repository name, the Git ref (branch, tag, commit SHA), the event that triggered the current run.</p>
<p>In your case, the information you require can be get from the <strong>github context</strong>. In your workflow, you also can print the whole context of the <strong>github context</strong> to the logs and view more details.</p>
<pre><code class="lang-yaml">- name: Print GitHub context
  run: echo "$GITHUB_CONTEXT"
  env:
    GITHUB_CONTEXT: ${{ toJson(github) }}
</code></pre>
<p>The expressions “<strong>github.event.pull_request.user.login</strong>” and “<strong>github.event.issue.user.login</strong>” are used to get the values of the corresponding properties from the <strong>github context</strong>.</p>

**Comment by zachliu** [comment]
<p>Great it works! Thank you!</p>
<p>When B comments on A’s PR, I can use <code>{{ github.event.issue.user.login }}</code> to get A’s username.</p>
<p>Now I am exploring <code>{{ toJson(github) }}</code> because I have another use case: when B <strong>merges</strong> A’s PR, I also want A’s username :grin:</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<blockquote>
<p>when B  <strong>merges</strong>  A’s PR,</p>
</blockquote>
<p>At this time, the event should be <strong>pull_request</strong>, and the activity type should be <strong>closed</strong> of the <strong>pull_request</strong> event. So, to get the the author of the PR, you need to use the expression “<strong>github.event.pull_request.user.login</strong>”.</p>
<br>
<blockquote>
<p>Great it works!</p>
</blockquote>
<p>Glad that the problems no longer bother you.<br>
If you think my above explanations is helpful to you to solve your questions, maybe you can mark it as the solution of this topic. So that other people who have the similar questions can notice this when they are looking for answers. Thanks.</p>

**Comment by zachliu** [comment]
<p>I have</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - master
</code></pre>
<p>but yeah, using <code>pull_request</code> + <code>closed</code> might be better</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/zachliu">@zachliu</a>,</p>
<p>If you want the workflow is triggered when the PR is merged, yes, using <strong>pull_request</strong> + <strong>closed</strong> is good.</p>
<pre><code class="lang-yaml">on:
  pull_request:
    types: [closed]
</code></pre>

**Comment by zachliu** [comment]
<p>Thank you so much!</p>
<p>Since I’m commenting/closing my own PRs all night, I couldn’t <strong>really</strong> verify the solution :sweat_smile: . I’ll double check tmr morning and mark the solution then.</p>
<p><strong>Update</strong>: now I can confirm everything works fine. Thank you! <a class="mention" href="/brightran">@brightran</a></p>
